packages/container-annotation/ContainerAnnotationManager.js

```diff
@@ -0,0 +1,164 @@
+import { Marker } from '../../model'
+
+class ContainerAnnotationManager {
+
+  constructor(context) {
+    if (!context.editorSession) {
+      throw new Error('EditorSession required.')
+    }
+
+    this.editorSession = context.editorSession
+    this.editorSession.onUpdate('document', this._onDocumentChanged, this)
+
+    this.doc = this.editorSession.getDocument()
+    this.context = Object.assign({}, context, {
+      // for convenienve we provide access to the doc directly
+      doc: this.doc
+    })
+
+    this._state = {
+      annotations: {},
+      containerFragments: {}
+    }
+
+    this.initialize()
+  }
+
+  dispose() {
+    this.editorSession.off(this)
+  }
+
+  initialize() {
+    this._computeAnnotations()
+    this._createAnnotations()
+    // HACK: we make commandStates dirty in order to trigger re-evaluation
+    this.editorSession._setDirty('commandStates')
+    this.editorSession.startFlow()
+  }
+
+  getAnnotationFragments(annoId) {
+    return this._state.containerFragments[annoId]
+  }
+
+  _onDocumentChanged(change) {
+    let dirtyNodes = []
+    let shouldUpdate = false
+
+    // For now we will run recomputation after any node creation
+    if(Object.keys(change.created).length > 0) shouldUpdate = true
+
+    // We will collect ids of updated nodes to update markers there
+    let updated = Object.keys(change.updated)
+    updated.forEach(prop => {
+      let nodeId = prop.split(',')[0]
+      let node = this.doc.get(nodeId)
+      if(node && node.isText()) {
+        if(dirtyNodes.indexOf(nodeId) === -1) {
+          dirtyNodes.push(nodeId)
+          // We will run recomputation and update if node with 
+          // conntainer annotation got updated
+          if(this._state.annotations[nodeId]) shouldUpdate = true
+        }
+      }
+    })
+
+    // Compute range for created/removed annos
+    if(dirtyNodes.length === 2) {
+      // TODO: We should get container via API
+      const container = this.doc.get('body')
```

Comment:
We need to fix this.
I need to take a closer look at this method. Maybe I have an idea.